bin/internal/last_engine_commit.sh

```diff
@@ -26,4 +26,37 @@ set -e
 
 FLUTTER_ROOT="$(dirname "$(dirname "$(dirname "${BASH_SOURCE[0]}")")")"
 
-git log -1 --pretty=format:%H -- "$(git rev-parse --show-toplevel)/DEPS" "$(git rev-parse --show-toplevel)/engine"
+# 1. Determine when we diverged from master, and prevent set -e from exiting.
+MERGE_BASE_COMMIT="$(git merge-base HEAD master || echo "")"
+
+# If we did not find a merge-base, fail
+if [[ -z "$MERGE_BASE_COMMIT" ]]; then
+  echo >&2 "Error: Could not determine a suitable engine commit."
+  echo >&2 "Current branch: $(git rev-parse --abbrev-ref HEAD)"
+  echo >&2 "Expected a different branch, from master"
+  exit 1
+fi
+
+# 2. Define and search history range to searhc within (unique to changes on this branch).
+HISTORY_RANGE="$MERGE_BASE_COMMIT..HEAD"
+ENGINE_COMMIT="$(git log -1 --pretty=format:%H --ancestry-path "$HISTORY_RANGE" -- "$(git rev-parse --show-toplevel)/DEPS" "$(git rev-parse --show-toplevel)/engine")"
+
+# 3. If no engine-related commit was found within the current branch's history, fallback to the first commit on this branch.
+if [[ -z "$ENGINE_COMMIT" ]]; then
+  # Find the oldest commit on HEAD that is *not* reachable from MERGE_BASE_COMMIT.
+  # This is the first commit *on this branch* after it diverged from 'master'.
+  ENGINE_COMMIT="$(git log --pretty=format:%H --reverse --ancestry-path "$MERGE_BASE_COMMIT"..HEAD | head -n 1)"
```

Comment:
non-blocking; you can use "git log -1" instead of "| head -n 1" or `Select-Object -First 1`